SafeExamBrowser.SystemComponents/VirtualMachineDetector.cs

```diff
@@ -97,69 +106,114 @@ private bool IsVirtualRegistry()
 
-					// reconstruct the systemInfo.biosInfo string
-					var biosInfo = (string) configKey.GetValue("BIOSVendor") + " " + (string) configKey.GetValue("BIOSVersion");
-					var manufacturer = (string) configKey.GetValue("SystemManufacturer");
-					var model = (string) configKey.GetValue("SystemProductName");
+				if (!success)
+					continue;
 
-					isVirtualMachine |= IsVirtualSystemInfo(biosInfo, manufacturer, model);
+				// reconstruct the systemInfo.biosInfo string
+				string biosInfo = $"{(string) biosVendor} {(string) biosVersion}";
 
-					// hardware information of profile throughout installation etc. 
-					var computerIdsKey = Microsoft.Win32.Registry.LocalMachine.OpenSubKey($"SYSTEM\\HardwareConfig\\{configId}\\ComputerIds");
+				isVirtualMachine |= IsVirtualSystemInfo(biosInfo, (string) systemManufacturer, (string) systemProductName);
 
-					if (computerIdsKey == null)
-					{
+				// check even more hardware information 
+				IEnumerable<string> computerIdNames;
+				var computerIdsKey = $"{hwConfigKey}\\ComputerIds";
+				if (!registry.TryGetNames(computerIdsKey, out computerIdNames))
+					continue;
+
+				foreach (var computerIdName in computerIdNames)
+				{
+					// collect computer hardware summary (e.g. manufacturer&version&sku&...)
+					object computerSummary;
+					if (!registry.TryRead(computerIdsKey, computerIdName, out computerSummary))
 						continue;
-					}
-
-					foreach (var computerId in computerIdsKey.GetSubKeyNames())
-					{
-						// e.g. manufacturer&version&sku&...
-						var computerSummary = (string) computerIdsKey.GetValue(computerId);
-						isVirtualMachine |= IsVirtualSystemInfo(computerSummary, computerSummary, computerSummary);
-					}
+
+					isVirtualMachine |= IsVirtualSystemInfo((string) computerSummary, (string) systemManufacturer, (string) systemProductName);
 				}
 			}
 
+			return isVirtualMachine;
+		}
+
+		/// <summary>
+		/// Scans (synced) device cache for hardware info of the current device.
+		/// </summary>
+		private bool IsVirtualRegistryDeviceCache()
+		{
+			bool isVirtualMachine = false;
+
+			// device cache contains hardware about other devices logged into as well, so lock onto this device in case an innocent VM was logged into.
+			// in the future, try to improve this check somehow since DeviceCache only gives ComputerName
+			var deviceName = System.Environment.GetEnvironmentVariable("COMPUTERNAME");
+
 			// check Windows timeline caches for current hardware config
-			var deviceCacheKey = Microsoft.Win32.Registry.CurrentUser.OpenSubKey($"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\TaskFlow\\DeviceCache");
-			var currHostname = System.Environment.GetEnvironmentVariable("COMPUTERNAME");
+			const string deviceCacheParentKey = "HKCU\\SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\TaskFlow\\DeviceCache";
+			IEnumerable<string> deviceCacheKeys;
+			bool has_dc_keys = registry.TryGetSubKeys(deviceCacheParentKey, out deviceCacheKeys);
```

Comment:
Always use camelCase for variable names, underscores are normally only used for CONSTANT_NAMES.